sklearn/preprocessing/imputation.py

```diff
@@ -374,3 +375,328 @@ def transform(self, X):
             X[coordinates] = values
 
         return X
+
+
+class KNNImputer(BaseEstimator, TransformerMixin):
+    """Imputation for completing missing values using k-Nearest Neighbors.
+
+    Each sample's missing values are imputed from up to "n_neighbors" nearest
+    neighbors found in the training set. Each missing feature is then
+    imputed as the average, either weighted or unweighted, of these neighbors
+    who have a value for it. Where all neighbors have that feature value
+    missing, the training set average for that feature is used for imputation.
+
+    Parameters
+    ----------
+    missing_values : integer or "NaN", optional (default = "NaN")
+        The placeholder for the missing values. All occurrences of
+        `missing_values` will be imputed. For missing values encoded as np.nan,
+        use the string value "NaN".
+
+    n_neighbors : int, optional (default = 5)
+        Maximum number of neighboring samples to use for imputation. When any
+        of the neighbors themselves have the feature value missing then the
+        remaining neighbors, if any, that have the feature value available are
+        used.
+
+    weights : str or callable, optional (default = "uniform")
+        Weight function used in prediction.  Possible values:
+
+        - 'uniform' : uniform weights.  All points in each neighborhood
+          are weighted equally.
+        - 'distance' : weight points by the inverse of their distance.
+          in this case, closer neighbors of a query point will have a
+          greater influence than neighbors which are further away.
+        - [callable] : a user-defined function which accepts an
+          array of distances, and returns an array of the same shape
+          containing the weights.
+
+    metric : str, optional (default = "masked_euclidean")
+        Distance metric for searching neighbors. Possible values:
```

Comment:
What would happen if the user provided a callable that implemented masked manhattan distance?

Reply:
Apocalypse? :) ... So it ends up calling NearestNeighbors().fit() which contains allow_nans = self.metric in _MASKED_METRICS which will return False and that means it will subsequently fail a check_array() test.

Reply:
So either this should change: the acceptance of Nans should be a function of the context rather than the metric, or remove the metric parameter, or clarify the documentation.

Reply:
Oh I thought we had (sort of) discussed this? You probably don't remember but I had asked if I could keep the parameter for now even though only a single option was available because I was planning to add a masked_manhattan once this was merged. Admittedly you had not explicitly "okayed" that but I was hoping the lack of a negative response meant a tacit approval :) -- was I mistaken there?

Reply:
ops missed the docs bit. what changes to the docs did you have in mind so that the parameter would be okay?

Reply:
I suppose the current docs are okay. It's just a bit weird :\

I don't particularly like how you won't be able to support a callable under the current design, now or in the future. I.e. this is a problem of how you determine whether to allow nans. Not to say that supporting callables will leave the code pretty. Although we could just minimise validation for callable metrics in `pairwise_distances` and that would solve that.

Reply:
I don't think I was very clear then. What I mean is that `allow_nans=metric == 'masked_euclidean'` is basically the current semantics. If we had `allow_nans=callable(metric) or metric == 'masked_euclidean'` is much more friendly for the future, and might even mean we can support alternative metrics now.